Merge from 3C

mattmccutchen-cci opened the pull request:
Hi!  As I trust @mwhicks1 has previously discussed with you, here is a pull request of Correct Computation's entire master branch with improvements to 3C.

As you can see, many files have been both renamed and heavily edited due to correctcomputation/checkedc-clang#299.  To make this PR easier to review, here are links to two diffs, which make up this PR when applied in sequence:
- [File renames](https://github.com/microsoft/checkedc-clang/compare/master...correctcomputation:file-renames-for-microsoft).
- [All other changes](https://github.com/correctcomputation/checkedc-clang/compare/file-renames-for-microsoft..master).  GitHub is not very good at displaying this, so you may prefer to run:
    ```
    git remote add -f correctcomputation https://github.com/correctcomputation/checkedc-clang
    git diff correctcomputation/{file-renames-for-microsoft,master}
    ```
    You can use the `--stat` option to `git diff` to help find changed non-3C-specific files.

Please disregard the old review comments from our repository that are showing up on this thread; we don't know why that is happening and haven't found a way to stop it.

mattmccutchen-cci commented:
This is ready for review.  Enjoy!

mgrang commented on clang/docs/checkedc/3C/CONTRIBUTING.md:
```diff
@@ -0,0 +1,36 @@
+# Contributing to 3C
```
Please format lines in this file to the 80 char limit.

mgrang commented on clang/docs/checkedc/3C/INSTALL.md:
```diff
@@ -0,0 +1,49 @@
+# Building 3C
```
Please format lines in this PR to the 80 char limit.

mgrang commented on README.md:
```diff
@@ -54,6 +54,10 @@ manual is
 
+## 3C: Semi-automated conversion of C code to Checked C
+
+This repository includes a tool called 3C that partially automates the conversion of C code to Checked C. Here is [general information about the 3C software](clang/docs/checkedc/3C/README.md), including development status and how to contribute. Here are the [usage instructions for the `3c` command-line tool](clang/tools/3c/README.md).
```
Please follow the 80 char line limit.

mgrang commented on clang/docs/checkedc/3C/README.md:
```diff
@@ -0,0 +1,11 @@
+# 3C: Semi-automated conversion of C code to Checked C
```
Please follow the 80 char line limit in this file.

mgrang commented on clang/tools/3c/README.md:
```diff
@@ -0,0 +1,65 @@
+# `3c`: Command-line tool for conversion of C code to Checked C
```
Please follow the 80 char line limit in this file.

mgrang commented on clang/tools/3c/utils/README.md:
```diff
@@ -0,0 +1,58 @@
+# Conversion Utilities
```
Please follow the 80 char line limit in this file.

mgrang commented on clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:
```diff
@@ -0,0 +1,929 @@
+//=--ArrayBoundsInferenceConsumer.cpp ----------------------------*- C++-*-===//
```
Please follow the 80 char line limit for lines in this file header.

mgrang commented on clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:
```diff
@@ -0,0 +1,929 @@
+  CVarSet ConsVar = CR.getExprConstraintVars(E);
+  const auto &EnvMap = Info.getConstraints().getVariables();
+  for (auto CurrCVar : ConsVar) {
+    if (needArrayBounds(CurrCVar, EnvMap) || needNTArrayBounds(CurrCVar, EnvMap))
```
80 char line limit.

mgrang commented on clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:
```diff
@@ -0,0 +1,929 @@
+static std::string getCalledFunctionName(const Expr *E) {
+  const CallExpr *CE = dyn_cast<CallExpr>(E);
+  assert(CE && "The provided expression should be a call expression.");
+  const FunctionDecl *CalleeDecl = dyn_cast<FunctionDecl>(CE->getCalleeDecl());
```
80 char line limit.

mgrang commented on clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:
```diff
@@ -0,0 +1,929 @@
+    // expression in which case if the type matches we use count bounds.
+    bool IsByteBound = true;
+    for (auto *TmpE : ArgVals) {
+      UnaryExprOrTypeTraitExpr *arg = dyn_cast<UnaryExprOrTypeTraitExpr>(TmpE);
```
80 char line limit.

mgrang commented on clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:
```diff
@@ -0,0 +1,929 @@
+        // tmp <- bounds(count)
+        // p->arr <- tmp
+        // TODO: This trick of introducing bounds for RHS expressions
+        //  could be useful in other places (e.g., &{1,2,3} would have bounds 3)
```
80 char line limit.

mgrang commented on clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:
```diff
@@ -0,0 +1,929 @@
+  if (FD->isThisDeclarationADefinition() && FD->hasBody()) {
+    auto &ABInfo = Info.getABoundsInfo();
+    auto &ABStats = ABInfo.getBStats();
+    const clang::Type *Ty = FD->getTypeSourceInfo()->getTypeLoc().getTypePtr();
```
80 char line limit.

mgrang commented:
LLVM follows a 80 char line limit. See https://llvm.org/docs/CodingStandards.html#source-code-width

Please make sure lines in this PR follow the 80 char line limit. Also please ensure all comments are terminated by a period.

mattmccutchen-cci commented:
I'm working on addressing the long lines in all 3C-specific files (not only the ones modified in this PR).  Reviewers, do you have any other comments in the meantime?

mgrang commented on clang/include/clang/3C/DeclRewriter.h:
```diff
@@ -0,0 +1,128 @@
+  public:
+    FieldFinder(GlobalVariableGroups &GVG) : GVG(GVG) { }
+
+    bool VisitFieldDecl(FieldDecl *FD);
```
I see that some functions use `functionName` while others use `FunctionName` as the naming convention. Could you please make the function cases uniform. Please refer to the LLVM Coding Standards at https://llvm.org/docs/CodingStandards.html

mattmccutchen-cci replied:
It turns out that this occurrence of `VisitFieldDecl` isn't our fault (it's a pseudo-override of `RecursiveASTVisitor::VisitFieldDecl` using the Curiously Recurring Template Pattern), but our code may well have other naming violations, and I'm working on setting up clang-tidy to find them all.

mattmccutchen-cci replied:
I've completed a clang-tidy pass, and this PR is now fully compliant with the implementation of the LLVM naming guideline in the `readability-identifier-naming` rule of clang-tidy version 11.0.0, except for documented suppressions. Please let me know if you have any concerns about our code violating aspects of the guideline that are not enforced by clang-tidy.

To expand on my previous comment: There are several exceptional scenarios in which we can't change names in our code because they have to match names that are not under our control. In some scenarios, clang-tidy doesn't recognize the coupling and would warn about the name, so we have a documented suppression comment. In other scenarios, clang-tidy version 11.0.0 (which we are using and this repository will soon be using!) knows not to warn, though the reason may not always be obvious to a person looking at the code. The scenarios I'm aware of that fall into this second category are:

1. Virtual overriding methods. In the future, we could ensure these are always obvious by enforcing the use of the `override` keyword using the `modernize-use-override`.
2. "Pseudo-overriding" methods that are referenced by name from a template and hide a method of the same name in a superclass that provides a default implementation. Methods named `Visit*` in subclasses of `RecursiveASTVisitor` generally fall into this category.

So if you see names in our code that look bad and don't have a documented suppression, you can either (a) run clang-tidy version 11.0.0, see that it doesn't warn, and trust that there's a reason for that; (b) research yourself which exceptional scenario applies; or (c) ask me. :slightly_smiling_face: 

We are mostly compliant with the other standard clang-tidy rules for the LLVM codebase too.

mattmccutchen-cci commented:
Hey, I noticed that it looks like someone on this repository (@kkjeer?) is working on upgrading to LLVM 11.  What are your plans for this?  When will it land on your master?  (Or is this already documented somewhere that I didn't see?)  It looks like you had to make some changes to 3C files, so there are likely to be some semantic (if not syntactic) conflicts with this PR (I haven't looked carefully); I'll be happy to help resolve them when the time comes.

It looks like you squashed the entire diff between your current master and its baseline LLVM revision into [a single commit](https://github.com/microsoft/checkedc-clang/commit/fed360dc13a706053d874226d0c1256493fb872f) on the LLVM 11 branch.  Compared to a normal merge, this will make it much more work to trace the history of pieces of code across the upgrade.  Did you see offsetting benefits of the squash, or problems with the merge?  If you'd be open to doing a merge instead but could use help with it, I'd be happy to help.

If you proceed with the squash, have you thought about how it will interact with this PR?  I assume that if LLVM 11 lands on master first, we'll archive the current history of the PR and replace it with a squash, and then any further additions we make to the PR would be on top of that squash.  If this PR lands first, then you would apply a squash of it with any necessary syntactic and semantic conflict resolutions (possibly with our help) to the LLVM 11 branch, as you would with anything else you merge to master while the LLVM 11 branch is pending.

mattmccutchen-cci commented:
Any more problems with the PR while I'm working on the naming conventions?

mattmccutchen-cci commented:
@mgrang FYI, I am still working toward fixing our names in bulk using clang-tidy; it may take me a few more days.  I know this is taking longer than we all would like.  I'm working only 10 hours per week for CCI and have had to investigate several clang-tidy bugs and had a few other CCI responsibilities to deal with as well.  I've added clang-tidy suppression comments for some things we don't want to rename (locally, not yet pushed), so I'm closer to reviewing the remaining renames proposed by clang-tidy to see if they are OK or there are other problems I need to address.

mgrang commented:
> @mgrang FYI, I am still working toward fixing our names in bulk using clang-tidy; it may take me a few more days. I know this is taking longer than we all would like. I'm working only 10 hours per week for CCI and have had to investigate several clang-tidy bugs and had a few other CCI responsibilities to deal with as well. I've added clang-tidy suppression comments for some things we don't want to rename (locally, not yet pushed), so I'm closer to reviewing the remaining renames proposed by clang-tidy to see if they are OK or there are other problems I need to address.

Adding @kkjeer since Katherine is working on the source upgrade. I guess it is still another week before we are ready to upgrade our sources.

mgrang commented:
Thanks for addressing the review comments. I ran our ADO validations on your PR: https://dev.azure.com/msresearch/CheckedC/_build/results?buildId=591077&view=results

I see a couple of warnings while building the lit tests:
```
/mnt/work/9/s/clang/lib/3C/AVarBoundsInfo.cpp:901:23: warning: unused variable ‘TKVar’ [-Wunused-variable]
           ProgramVar *TKVar = BI->getProgramVar(TK);

/mnt/work/9/s/clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:616:9: warning: unused variable ‘ABoundsInfo’ [-Wunused-variable]
   auto &ABoundsInfo = Info.getABoundsInfo();
```

mattmccutchen-cci commented:
> Thanks for addressing the review comments. I ran our ADO validations on your PR: https://dev.azure.com/msresearch/CheckedC/_build/results?buildId=591077&view=results

When I try to view that link, I'm asked to log in to a Microsoft account.  Can I use any Microsoft account, do I need some permission, or will you just tell me anything I need to know from there?

> I see a couple of warnings while building the lit tests:
> 
> ```
> /mnt/work/9/s/clang/lib/3C/AVarBoundsInfo.cpp:901:23: warning: unused variable ‘TKVar’ [-Wunused-variable]
>            ProgramVar *TKVar = BI->getProgramVar(TK);
> 
> /mnt/work/9/s/clang/lib/3C/ArrayBoundsInferenceConsumer.cpp:616:9: warning: unused variable ‘ABoundsInfo’ [-Wunused-variable]
>    auto &ABoundsInfo = Info.getABoundsInfo();
> ```

Fixed.

mgrang commented:
I don't see any other issues with the ADO run. The changes look good to me.

@dtarditi Could you take a final look at the PR? Also do we have a plan for the 3C team to run their own ADO validations for their PRs in future?

dtarditi commented on clang/docs/checkedc/3C/CONTRIBUTING.md:
```diff
@@ -0,0 +1,92 @@
+agreement (CLA) process at this time, but we may set one up and
+require you to complete it before we accept your contribution. Also be
+aware that we need to keep 5C working, so you may have to wait for us
+to address 5C-specific problems arising from your 3C pull request
```
You might to provide a sentence on 5C here.  At first, I thought there was a typo and then I read further.

mattmccutchen-cci replied:
Added a link to the section of our readme that describes 5C.

dtarditi commented on clang/docs/checkedc/3C/INSTALL.md:
```diff
@@ -0,0 +1,106 @@
+## CR/LF issues
+
+There is a [known
+problem](https://github.com/correctcomputation/checkedc-clang/issues/317)
```
We have a PR out to fix this: see #950. We ran into similar issues and tracked it down to .gitattributes as well.  This was introduced in an attempt to keep people from checking in files from Windows machines with accidental line feeds/carriage returns.  Git does not handle this well.

mattmccutchen-cci replied:
Yes, I noticed the commit in connection with the LLVM upgrade but not the PR; thanks for the pointer.  I'm resolving this thread because it's not a problem with this PR.  I'll update our documentation as soon as feasible after you merge the fix.

mattmccutchen-cci replied:
Oops, I see #950 was already merged.  Still, I want to get this PR merged ASAP and I hope it's OK if I wait to update our documentation until I have time to test your fix; that will be tracked by correctcomputation/checkedc-clang#317.

mattmccutchen-cci replied:
By the way, I'm CCI's resident git expert (as you might have gathered) and would be happy to help you fix any line-ending-related problems in your workflows that remain after #950, subject to my other priorities.

dtarditi commented on clang/include/clang/3C/3CGlobalOptions.h:
```diff
@@ -0,0 +1,37 @@
+extern bool Verbose;
+extern bool DumpIntermediate;
+extern bool HandleVARARGS;
+extern bool SeperateMultipleFuncDecls;
```
There's a typo here: seperate->separate.  You might consider down the road alphabetizing these variables and adding comments.

mattmccutchen-cci replied:
Filed correctcomputation/checkedc-clang#344.

dtarditi reviewed (approved):
Looks great, thank you for the contributions again! I have a few minor suggestions for improvements for the future. @mgrang , could you merge this?

mattmccutchen-cci commented:
> Looks great, thank you for the contributions again! I have a few minor suggestions for improvements for the future. @mgrang , could you merge this?

Great!  If you can hang on an hour so until after my other meeting, I'll slip a fix for the 5C thing into this PR.

mattmccutchen-cci commented:
I merged from your master and our tests pass.  Ready to go!

mgrang commented:
This is breaking our Windows x64 builds with the error:
`D:\work\13\s\clang\lib\3C\RewriteUtils.cpp : fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj`

It seems we had a fix for this on master which seems to have been removed in this commit. I have a fix to add it back: https://github.com/microsoft/checkedc-clang/pull/951

mattmccutchen-cci commented:
> It seems we had a fix for this on master which seems to have been removed in this commit.

Sorry about that.  I'm researching how it happened.  It looks like it was in a period before I joined CCI when our git history was a mess.

mgrang commented:
Matt, several 3C unit tests are failing on the X86 Window build.
```
Failing Tests (9):
    Clang :: 3C/3d-allocation.c
    Clang :: 3C/basic.c
    Clang :: 3C/definedType.c
    Clang :: 3C/extstructfields.c
    Clang :: 3C/graphs.c
    Clang :: 3C/graphs2.c
    Clang :: 3C/linkedlist.c
    Clang :: 3C/multivardecls.c
    Clang :: 3C/root_cause.c
```

mgrang commented:
[Unit Test Failures.txt](https://github.com/microsoft/checkedc-clang/files/5645377/Unit.Test.Failures.txt)

mattmccutchen-cci commented:
> Matt, several 3C unit tests are failing on the X86 Window build.

If the set of tests that fail is consistent, can you try marking them `UNSUPPORTED: windows` (see [documentation](https://llvm.org/docs/TestingGuide.html#constraining-test-execution)) until we have time to come up with a real fix?  I'd send you a PR, but it's probably faster if you are the one to iterate on getting the syntax for this right.

Next time, whatever tests you plan to run on your master after merging, can you please run them on our PR branch before merging?  (I have been ensuring that the PR branch is up to date with your master so the content of the merge will equal the PR branch; you can also check that yourself.)  We can't drop everything for a rush fix to a bunch of unexpected failures on your master branch.  There was some discussion of us setting up a Windows test environment, but I don't know if we can ever be certain that it will exactly match yours.  This time, if the problems are bad enough, it's fine with me if you revert the whole thing and we send another PR once we have the fix.

mgrang commented:
I have a PR to disable the failing unit tests: https://github.com/microsoft/checkedc-clang/pull/952
I have also filed an issue to track this: https://github.com/correctcomputation/checkedc-clang/issues/345

mattmccutchen-cci commented:
> > It seems we had a fix for [the `/bigobj` issue] on master which seems to have been removed in this commit.
>
> Sorry about that. I'm researching how it happened; it looks like it was in a period before I joined CCI when our git history was a mess.

My conclusions, finally:

I confirmed that the reversion of both fixes from #917 (`/bigobj` and `#include <cctype>`) and a third erroneous change we discovered in our own work all occurred in our merge in 4c4eed3c94dff03eb6ab07ee745e2ca6ce4cbea4, and the root cause of all three errors was the fact that #891 was squash-merged.  In this merge, the left side included the source commit of #891 (740d75881f276e68a5910b24faa5f5dfdb63d501) and not the squash merge of #891 (3f681137e9eba213218cdaa947a4336135edf111), while the opposite was true of the right side.

If #891 had been merged with a normal merge commit, then Git would have been able to follow the edge from the merge commit to the source commit, the merge base for 4c4eed3c94dff03eb6ab07ee745e2ca6ce4cbea4 would have been 7fc929240bcd87301c4f6d2883c5165a384e54aa, and the automated textual merge in 4c4eed3c94dff03eb6ab07ee745e2ca6ce4cbea4 would have done the right thing in all three cases.  But since Git was unaware of the relationship between the squash merge and the source commit, it went all the way back to 999bba65b77dc020185d441df72bd001896783da and 253189545698f340a9633c5c3ea467ec1199884b as merge bases.  Each change between (999bba65b77dc020185d441df72bd001896783da / 253189545698f340a9633c5c3ea467ec1199884b) and 7fc929240bcd87301c4f6d2883c5165a384e54aa appeared in its original form on the left side of 4c4eed3c94dff03eb6ab07ee745e2ca6ce4cbea4 and as part of the squash of #891 on the right side of 4c4eed3c94dff03eb6ab07ee745e2ca6ce4cbea4.  I'll call such a change a "duplicated" change.

In the `/bigobj` and `#include <cctype>` cases, we had made a duplicated change.  The left side of 4c4eed3c94dff03eb6ab07ee745e2ca6ce4cbea4 had the original change, while the right side had the other copy of the change _plus_ the subsequent changes in #917.  This generated a conflict, and our manual resolution of this conflict was incorrect.  In the third case, the sequence of events was different: we made a duplicated change and then reverted it after submitting #891, so the left side saw no net change while the right side saw the squashed copy of the forward change.  The automated textual merge silently took the forward change, which was not what we intended.

These three errors spooked me enough that I conducted an investigation to try to find any more latent errors all at once rather than having them cause problems for us one at a time.  Specifically, I used a custom tool to analyze every merge we ever performed, checking whether the outcome might have been different if all the 3C PRs from CCI's repository to Microsoft's had been merged normally instead of squash-merged.  The tool identified several merges that were potentially incorrect and proposed patches that could have corrected them at the time they were originally performed.  In a few cases, it was clear to me that a patch addressed an error that still exists and the patch should be applied now (#962 and correctcomputation/checkedc-clang#364).  But a quick look at the rest of the patches suggested they were probably no longer relevant: we may have later rewritten the affected code, scrutinized it thoroughly enough that we would have corrected any problems we could conceivably care about, or even noticed and corrected the specific merge error in a later commit (the tool did not check for this).  I didn't think it was worth spending the effort to research them more thoroughly on the off chance that there might be something we might want.

To prevent future problems of this nature, Microsoft's agreement to perform normal rather than squash merges of future omnibus 3C PRs (contingent on us keeping the commit sequences of those PRs reasonably clean) should be sufficient.  (Since squash merging still seems to be the habit, e.g., for #956, I'll be sure to remind you multiple times!)  I also plan to keep an eye on things for the next few months, and if your team does squash-merge an omnibus 3C PR by mistake, there is a step I can take to prevent it from causing later merging mistakes (as I did for #930 in b23b287729cfd1c7f5709043469de808af9b1df4).
